services/web-app/utils/mdx-image-optimizer-plugin.mjs

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright IBM Corp. 2022, 2022
+ *
+ * This source code is licensed under the Apache-2.0 license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+import path from 'path'
+import { getPlaiceholder } from 'plaiceholder'
+import { remove } from 'unist-util-remove'
+import { visit } from 'unist-util-visit'
+
+/**
+ * Determines if two urls belong to the same github repo
+ *
+ * @param {string} url1 first repo to compare
+ * @param {stirng} url2 second repo to compare
+ * @returns {boolean} whether the two urls belong to the same github repo or not
+ */
+const urlsBelongToTheSameRepo = (url1, url2) => {
+  // first four chunks of urls must be equal:
+  // https://github.com/org/repo/*
+  const urlObj1 = new URL(url1)
+  const urlObj2 = new URL(url2)
+
+  // check that hostname matches
+  if (urlObj1.hostname !== urlObj2.hostname) {
+    return false
+  }
+
+  // check that org and repo match
+  return (
+    JSON.stringify(urlObj1.pathname.split('/', 3)) ===
+    JSON.stringify(urlObj2.pathname.split('/', 3))
+  )
+}
+
+/**
+ * finds all image nodes in tree and:
+ * 1 - appends dirPath to image url if url is relative path
+ * (assumes dirPath is url of base of a github repo in which the image lives)
+ * 2 - Removes image nodes if relative url attempts to modify base url
+ * 3 - converts image node to next/Image component
+ * 4 - Adds height, width and placeholder properties to image
+ *
+ * @param {string} dirPath path where mdx source comes from (expected to be github repo url)
+ * @param {object} tree AST: mdx source
+ * @returns undefined
+ */
+const optimizeTreeImages = async (dirPath, tree) => {
+  const matches = []
+  visit(
+    tree,
+    (node) => node.type === 'image' || (node.type === 'mdxJsxFlowElement' && node.name === 'Image'),
```

Comment:
It might be clearer to use a switch (or something similar) here to separate the two paths that are currently intertwined: one for if a markdown image is encountered and another if an Image mdxJsxFlowElement is encountered.

This would bring the logic more in line with typical XML parsers, whereby each unique type of element that is encountered has its own handler during the parse/visit.